File: Phase01/main/src/Token.java
public class Token implements Comparable<Token> {

    private String key;

    public Token(String key) {
        this.key = normalize(key);
    }

    public String getKey() {
        return key;
    }

    public void setKey(String newKey) {
        this.key = newKey;
    }

    private String normalize(String key) {
        String normalized = key;
        normalized = key.toLowerCase();
        return normalized;
    }

    @Override
    public String toString() {
        return "Token(" + this.key + ")";
    }

    @Override
    public boolean equals(Object other) {
        if (this == other)
            return true;
        if (!(other instanceof Token))
            return false;
        Token that = (Token) other;
        return this.key.equals(that.key);
    }

    @Override
    public int hashCode() {
        return key.hashCode();
    }

    @Override
    public int compareTo(Token that) {
        // returns -1 if "this" object is less than "that" object
        // returns 0 if they are equal
        // returns 1 if "this" object is greater than "that" object
        return this.key.compareTo(that.key);
    }

}

File: GoogleSharp/GoogleSharp/src/core/structures/Token.cs
// Standard Library
using System;

namespace GoogleSharp.Src.Core.Structures {
    public class Token : IComparable {
        private string key;

        public Token (string key) {
            Key = key;
        }

        public string Key {
            get {
                return key;
            }
            set {
                key = Normalize (value);
            }
        }

        public override bool Equals (object obj) {
            if (obj is Token other)
                return Key == other.Key;
            return false;
        }

        public override int GetHashCode () {
            return Key.GetHashCode ();
        }

        public override string ToString () {
            return $"Token({key})";
        }

        public int CompareTo (object obj) {
            // returns -1 if "this" object is less than "that" object
            // returns 0 if they are equal
            // returns 1 if "this" object is greater than "that" object
            if (obj is Token that)
                return Key.CompareTo (that.Key);
            throw new ArgumentException ("Object is not a token !");
        }

        private static string Normalize (string key) {
            string normalized = key;
            normalized = normalized.ToLower ();
            return normalized;
        }
    }
}
